src/Inputs/Slider.tsx

```diff
@@ -1,79 +1,440 @@
-import React from 'react';
-import styled from 'styled-components';
-import { LabelLayout, LabelLayoutProps } from '../Fragments';
+import React, {
+    useState,
+    useLayoutEffect,
+    useRef,
+    RefObject,
+    MouseEvent,
+    useMemo,
+} from 'react';
+import styled, { DefaultTheme } from 'styled-components';
+import { LabelLayout, LabelLayoutProps } from '@Layouts';
+import { Popup } from '../Containers/Popup';
+
+export interface MarkProps {
+    key: number;
+    mark: string;
+}
+
+export interface ResultValues {
+    lowValue: number;
+    highValue: number;
+}
 
 export interface SliderProps extends LabelLayoutProps {
+    result?: ResultValues;
     max?: number;
     min?: number;
     step?: number;
     disabled?: boolean;
-    value?: number;
-    trackColor?: string;
-    buttonColor?: string;
+    hasTwoInputs?: boolean;
+    marks?: MarkProps[];
+    onChange?: Function;
+    valueFinish?: number;
+    valueStart?: number;
+    hasRail?: boolean;
+    left?: number;
+    right?: number;
+    hasPopup?: boolean;
+    popupleft?: number;
+    popuptop?: number;
+    popupwidth?: number;
+    popupheight?: number;
+    theme?: DefaultTheme;
 }
 
 export const Slider: React.FunctionComponent<SliderProps> = ({
+    disabled = false,
+    marks,
+    min = 0,
+    max = 100,
+    hasRail,
+    hasTwoInputs = false,
+    valueFinish = max,
+    valueStart = min,
+    hasPopup,
+    step = 1,
+    onChange = (): void => {},
+    popupleft = -17,
+    popuptop = -51,
+    popupwidth = 'auto',
+    popupheight = 20,
+    theme,
     ...props
-}): React.ReactElement => (
-    <LabelLayout {...props}>
-        <SliderElement {...props} type="range" />
-    </LabelLayout>
-);
-
-const SliderElement = styled.input<SliderProps>`
-    -webkit-appearance: none;
-    margin: 18px 0;
-    width: 100%;
-    // styles for slider adapted from: https://css-tricks.com/styling-cross-browser-compatible-range-inputs-css/
-    &:disabled {
-        cursor: not-allowed;
-        opacity: 0.6;
-    }
-    &:focus {
-        outline: none;
-    }
-    &::-webkit-slider-runnable-track {
-        width: 100%;
-        height: 6px;
-        cursor: pointer;
-        animate: 0.2s;
-        box-shadow: 1px 1px 1px #000000, 0px 0px 1px #0d0d0d;
-        background: #3071a9;
-        border-radius: 10px;
-        border: 0.2px solid ${(props): string => props.trackColor || '#010101'};
+}): React.ReactElement => {
+    // DOM Elements
+    const bar = useRef() as RefObject<HTMLDivElement>;
```

Comment:
const bar= useRef<HTMLDivElement>();
Can you do this instead?

Reply:
I'll get typescript errors whenever im using the props of bar

Reply:
I mean
useRef< HTMLDivElement >()

Reply:
I'll get ts errors that mutablerefobject<htmldivelement/undefined> is not assignable to type reconnect<htmldivelement> and ((instance:htmldivelement/null)=>voidect.

Reply:
if you read the error message it says instance has to be null or the htmldivelement.

Reply:
you have to define the ref as null as the ref must either be the element object or null, not undefined
![image](https://user-images.githubusercontent.com/47584841/82922953-43ab8f80-9f48-11ea-828b-96197e3d3f7e.png)